client/my-sites/checkout/cart/cart-free-user-plan-upsell.jsx

```diff
@@ -0,0 +1,175 @@
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+import React from 'react';
+import { connect } from 'react-redux';
+import { localize } from 'i18n-calypso';
+import { find } from 'lodash';
+import formatCurrency from '@automattic/format-currency';
+
+/**
+ * Internal dependencies
+ */
+import { Button } from '@automattic/components';
+import { getSelectedSite } from 'state/ui/selectors';
+import { siteHasPaidPlan } from 'signup/steps/site-picker/site-picker-submit';
+import { currentUserHasFlag, getCurrentUser } from 'state/current-user/selectors';
+import { PLAN_UPSELL_FOR_FREE_USERS } from 'state/current-user/constants';
+import {
+	getAllCartItems,
+	hasDomainRegistration,
+	hasPlan,
+	planItem,
+} from 'lib/cart-values/cart-items';
+import { addItem } from 'lib/cart/actions';
+import SectionHeader from 'components/section-header';
+import { PLAN_PERSONAL } from 'lib/plans/constants';
+import { isRequestingSitePlans } from 'state/sites/plans/selectors';
+import { isRequestingPlans } from 'state/plans/selectors';
+import { getPlan } from 'lib/plans';
+import { getPlanPrice } from 'state/products-list/selectors';
+import { isDomainRegistration } from 'lib/products-values';
+import TrackComponentView from 'lib/analytics/track-component-view';
+import { recordTracksEvent } from 'state/analytics/actions';
+
+class CartFreeUserPlanUpsell extends React.Component {
+	static propTypes = {
+		cart: PropTypes.object,
+		selectedSite: PropTypes.oneOfType( [ PropTypes.object, PropTypes.bool ] ),
+		hasPaidPlan: PropTypes.bool,
+		hasPlanInCart: PropTypes.bool,
+		isPlansListFetching: PropTypes.bool,
+		isRegisteringDomain: PropTypes.bool,
+		isSitePlansListFetching: PropTypes.bool,
+		personalPlan: PropTypes.oneOfType( [ PropTypes.object, PropTypes.bool ] ),
+		planPrice: PropTypes.oneOfType( [ PropTypes.number, PropTypes.bool ] ),
+		showPlanUpsell: PropTypes.bool,
+		translate: PropTypes.func.isRequired,
+	};
+
+	isLoading() {
+		const { cart } = this.props;
+		const isLoadingCart = ! cart.hasLoadedFromServer || cart.hasPendingServerUpdates;
+		const isLoadingPlans = this.props.isPlansListFetching;
+		const isLoadingSitePlans = this.props.isSitePlansListFetching;
+		return isLoadingCart || isLoadingPlans || isLoadingSitePlans;
+	}
+
+	getUpgradeText() {
+		const { cart, planPrice, translate } = this.props;
+		const firstDomain = find( getAllCartItems( cart ), isDomainRegistration );
```

Comment:
I think we already have a helper that achieves the same: https://github.com/Automattic/wp-calypso/blob/fe44257b9e3bbf6aff1692ef5601eda1d7588d4d/client/lib/cart-values/cart-items.js#L841
Also, it's possible to have more than one domain registration in cart, so be sure to take that into account :)

Reply:
I refactored this part of the code. And yes, that is being handled, since the first domain will be the discounted one if a plan gets added :)